analyzers/src/SonarAnalyzer.CSharp/Rules/Hotspots/PubliclyWritableDirectories.cs

```diff
@@ -0,0 +1,54 @@
+﻿/*
+ * SonarAnalyzer for .NET
+ * Copyright (C) 2015-2021 SonarSource SA
+ * mailto: contact AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using System;
+using System.Linq;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Diagnostics;
+using SonarAnalyzer.Common;
+using SonarAnalyzer.Helpers;
+
+namespace SonarAnalyzer.Rules.CSharp
+{
+    [DiagnosticAnalyzer(LanguageNames.CSharp)]
+    [Rule(DiagnosticId)]
+    public sealed class PubliclyWritableDirectories : PubliclyWritableDirectoriesBase<SyntaxKind, InvocationExpressionSyntax>
+    {
+        protected override ILanguageFacade<SyntaxKind> Language => CSharpFacade.Instance;
+
+        public PubliclyWritableDirectories() : this(AnalyzerConfiguration.Hotspot) { }
+
+        internal PubliclyWritableDirectories(IAnalyzerConfiguration configuration) : base(configuration, RspecStrings.ResourceManager) { }
+
+        private protected override bool IsGetTempPathAssignment(InvocationExpressionSyntax invocationExpression, KnownType type, string methodName, SemanticModel semanticModel) =>
+            invocationExpression.IsMethodInvocation(type, methodName, semanticModel)
+            && (invocationExpression.Parent is EqualsValueClauseSyntax
+                || invocationExpression.Parent is AssignmentExpressionSyntax
+                || invocationExpression.Parent is ArrowExpressionClauseSyntax
+                || invocationExpression.Parent is ReturnStatementSyntax);
+
+        private protected override bool IsInsecureEnvironmentVariableRetrieval(InvocationExpressionSyntax invocation, KnownType type, string methodName, SemanticModel semanticModel) =>
+            invocation.IsMethodInvocation(type, methodName, semanticModel)
+            && invocation.ArgumentList?.Arguments.FirstOrDefault() is { } firstArgument
+            && InsecureEnvironmentVariables.Any(x => x.Equals(firstArgument.Expression?.GetStringValue(), StringComparison.OrdinalIgnoreCase));
```

Comment:
It's a hotspot and most of the hotspots uses `.FindStringConstant` to also detect value propagation. Should we use that instead here as well?

We can handle the interpolated string separately.

Reply:
I would say we do not need that as whether propagation is used or not, we will raise at the location wherever the string literal is used?

Reply:
I don't understand :) Raising at the usage location is fine in all cases. Difference is if we're able to detect this
```
var name="%temp%";
Environment.GetEnvironmentVariable(name);
```

Reply:
We register for StringLiteralExpression and InterpolatedStringExpression so we do not care about value propagation.